### DISCUSSION TITLE
error: could not find native static library `fltk_gl`, perhaps an -L flag is missing?

### DISCUSSION THREAD

**Discussion by salmmanfred** [fltk-rs/fltk-rs#944] [Q&A]
how do i fix this fltk_gl is downloaded on my machine

**Comment by MoAlyousef** [comment]
It's built automatically by FLTK if the system has opengl, if it doesn't it won't be built. 
If you're using the enable-glwindow or fltk-bundled features, this assumes present and working opengl drivers and libs. Otherwise, you'll need to remove the flag from your Cargo.toml, or install the correct opengl libs.

**Reply by MoAlyousef** [reply]
Did the project build prior to a certain update of fltk-rs?
What operating system are you building for?

**Reply by salmmanfred** [reply]
whats the download command for GLUT on mingw msys?

**Reply by MoAlyousef** [reply]
Depending on your architecture:
pacman -S mingw-w64-x86_64-freeglut
pacman -S mingw-w64-i686-freeglut

However mingw usually uses the systems install of OpenGL which is usually at C:\Windows\System32\.
You should check if you have opengl32.dll and glu32.dll inside that directory.

**Reply by salmmanfred** [reply]
i have those

but the code refuses to compile still

**Reply by MoAlyousef** [reply]
What's in your Cargo.toml?
Please share the outputs of:
`rustup show`
and 
`cargo clean && cargo build -vv`

**Reply by salmmanfred** [reply]
thing is it builds fine on one of my computers but the other it just wont

cargo build -vv
Caused by:
  process didn't exit successfully: `set CARGO=\\?\C:\Users\ManfredMårtensson\.rustup\toolchains\stable-x86_64-pc-windows-msvc\bin\cargo.exe&& set CARGO_CRATE_NAME=fltk_sys&& set CARGO_MANIFEST_DIR=C:\Users\ManfredMårtensson\.cargo\registry\src\github.com-1ecc6299db9ec823\fltk-sys-1.1.11&& set CARGO_PKG_AUTHORS="MoAlyousef <mohammed.alyousef@neurosrg.com>"&& set CARGO_PKG_DESCRIPTION="Rust bindings for the FLTK GUI library"&& set CARGO_PKG_HOMEPAGE=""&& set CARGO_PKG_LICENSE=MIT&& set CARGO_PKG_LICENSE_FILE=""&& set CARGO_PKG_NAME=fltk-sys&& set CARGO_PKG_REPOSITORY=https://github.com/fltk-rs/fltk-rs&& set CARGO_PKG_VERSION=1.1.11&& set CARGO_PKG_VERSION_MAJOR=1&& set CARGO_PKG_VERSION_MINOR=1&& set CARGO_PKG_VERSION_PATCH=11&& set CARGO_PKG_VERSION_PRE=""&& set OUT_DIR=C:\Users\ManfredMårtensson\Documents\prog\CDCS\target\debug\build\fltk-sys-f08b071afa9c8c23\out&& set PATH="C:\Users\ManfredMårtensson\Documents\prog\CDCS\target\debug\deps;C:\Users\ManfredMårtensson\.rustup\toolchains\stable-x86_64-pc-windows-msvc\bin;C:\Users\ManfredMårtensson\.cargo\bin;C:\Users\ManfredMårtensson\.rustup\toolchains\stable-x86_64-pc-windows-msvc\bin;C:\windows\system32;C:\windows;C:\windows\System32\Wbem;C:\windows\System32\WindowsPowerShell\v1.0\;C:\windows\System32\OpenSSH\;C:\Program Files\dotnet\;C:\Program Files (x86)\dotnet\;C:\Program Files\Git\cmd;C:\Users\ManfredMårtensson\.cargo\bin;C:\Users\ManfredMårtensson\AppData\Local\Microsoft\WindowsApps;C:\Users\ManfredMårtensson\AppData\Local\Programs\Microsoft VS Code\bin;C:\Users\ManfredMårtensson\.dotnet\tools;C:\Users\ManfredMårtensson\AppData\Local\GitHubDesktop\bin;C:\Program Files\CMake\bin;C:\Program Files\Git\bin;"&& rustc --crate-name fltk_sys --edition=2018 C:\Users\ManfredMårtensson\.cargo\registry\src\github.com-1ecc6299db9ec823\fltk-sys-1.1.11\src/lib.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts --crate-type lib --emit=dep-info,metadata,link -C embed-bitcode=no -C debuginfo=2 --cfg "feature=\"default\"" --cfg "feature=\"enable-glwindow\"" --cfg "feature=\"fltk-bundled\"" -C metadata=f7afbb5b2a136ecb -C extra-filename=-f7afbb5b2a136ecb --out-dir C:\Users\ManfredMårtensson\Documents\prog\CDCS\target\debug\deps -L dependency=C:\Users\ManfredMårtensson\Documents\prog\CDCS\target\debug\deps --extern libc=C:\Users\ManfredMårtensson\Documents\prog\CDCS\target\debug\deps\liblibc-592f50b186c5d268.rmeta --cap-lints warn -L native=C:\Users\ManfredMårtensson\Documents\prog\CDCS\target\debug\build\fltk-sys-f08b071afa9c8c23\out\build -L native=C:\Users\ManfredMårtensson\Documents\prog\CDCS\target\debug\build\fltk-sys-f08b071afa9c8c23\out\build\Release -L native=C:\Users\ManfredMårtensson\Documents\prog\CDCS\target\debug\build\fltk-sys-f08b071afa9c8c23\out\lib -L native=C:\Users\ManfredMårtensson\Documents\prog\CDCS\target\debug\build\fltk-sys-f08b071afa9c8c23\out\lib64 -L native=C:\Users\ManfredMårtensson\Documents\prog\CDCS\target\debug\build\fltk-sys-f08b071afa9c8c23\out\lib\Release -L native=C:\Users\ManfredMårtensson\Documents\prog\CDCS\target\debug\build\fltk-sys-f08b071afa9c8c23\out\lib64\Release -l static=cfltk -l static=fltk -l static=fltk_images -l static=fltk_png -l static=fltk_jpeg -l static=fltk_z -l static=fltk_gl -l dylib=opengl32 -l dylib=glu32 -l dylib=ws2_32 -l dylib=comctl32 -l dylib=gdi32 -l dylib=oleaut32 -l dylib=ole32 -l dylib=uuid -l dylib=shell32 -l dylib=advapi32 -l dylib=comdlg32 -l dylib=winspool -l dylib=user32 -l dylib=kernel32 -l dylib=odbc32 -l dylib=gdiplus` (exit code: 1)

**Reply by MoAlyousef** [reply]
Can you also show the `rustup show` output? Also is that the whole output of `cargo clean && cargo build -vv` ?

You seem to be using the msvc toolchain, so you can check here to see if everything is installed properly:
https://github.com/fltk-rs/fltk-rs/wiki/Setup#windows-msvc-toolchain

If everything is installed, the other computer might have OpenGL driver issues, which you might need to install from the vendor. Something like:
https://developer.nvidia.com/opengl-driver
for my machine

**Reply by salmmanfred** [reply]
what about intel?

**Reply by MoAlyousef** [reply]
https://www.intel.com/content/www/us/en/support/articles/000005629/graphics.html

**Reply by salmmanfred** [reply]
i have the latest drivers

**Reply by tfinnegan937** [reply]
Hey, I hate to ping an old ticket, but I am having this same issue. Latest nvidia drivers, MSVC toolchain with the C++ build tools, Windows 10 and 11 SDKs, etc. This is on Windows 11.

Attempting to build with the following dependency:

`fltk = { version = "^1.2", features = ["fltk-bundled", "enable-glwindow"] }`

produces

`error: could not find native static library `fltk_gl`, perhaps an -L flag is missing?`

Here is the output of my rustup show call:

`Default host: x86_64-pc-windows-msvc
rustup home:  C:\Users\Tim\.rustup

stable-x86_64-pc-windows-msvc (default)
rustc 1.69.0 (84c898d65 2023-04-16)
`

Any idea what I could be missing?

**Reply by tfinnegan937** [reply]
Nevermind! My issue appears to have been simple. I removed fltk-bundled from my feature list, installed cmake, and added cmake to my path. I then rebooted my machine. enable-glwindow now works.
